.evergreen/csfle/activate-kmstlsvenv.sh

```diff
@@ -5,24 +5,93 @@
 # Usage:
 #   . ./activate-kmstlsvenv.sh
 #
-# This file creates and activates the kmstlsvenv virtual environment in the
+# This file creates and/or activates the kmstlsvenv virtual environment in the
 # current working directory. This file must be invoked from within the
 # .evergreen/csfle directory in the Drivers Evergreen Tools repository.
+#
+# If a kmstlsvenv virtual environment already exists, it will be activated and
+# no further action will be taken. If a kmstlsvenv virtual environment must be
+# created, required packages will also be installed.
+
+# If an error occurs during creation, activation, or installation of packages,
+# the kmstlsvenv virtual environment will be deactivated and activate_kmstlsvenv
+# will return a non-zero value.
 
 # Automatically invoked by activate-kmstlsvenv.sh.
 activate_kmstlsvenv() {
   # shellcheck source=.evergreen/venv-utils.sh
   . ../venv-utils.sh || return
 
   if [[ -d kmstlsvenv ]]; then
-    venvactivate kmstlsvenv
+    venvactivate kmstlsvenv || return
   else
     # shellcheck source=.evergreen/find-python3.sh
     . ../find-python3.sh || return
 
     venvcreate "$(find_python3)" kmstlsvenv || return
 
-    CRYPTOGRAPHY_DONT_BUILD_RUST=1 python -m pip install --upgrade boto3~=1.19 cryptography~=3.4.8 pykmip~=0.10.0
+    local packages=(
+      "boto3~=1.19.0"
+      "pykmip~=0.10.0"
+    )
+
+    if [[ "$OSTYPE" == darwin16 && "$HOSTTYPE" == x86_64 ]]; then
+      # Avoid `error: thread-local storage is not supported for the current
+      # target` on macos-1012.
+      packages+=("greenlet<2.0")
+    fi
+
+    if [[ "$OSTYPE" == cygwin && "$HOSTTYPE" == x86_64 ]]; then
+      local -r windows_os_name="$(systeminfo.exe /FO LIST | perl -lne 'print $1 if m/^OS Name:\s+(.*)$/' || true)"
+
+      if [[ "$windows_os_name" =~ 2016 ]]; then
+        # Avoid `RuntimeError: Could not determine home directory.` on
+        # windows-64-2016. See BUILD-16233.
+        python -m pip install -U "setuptools<65.0" || {
+          local -r ret="$?"
+          deactivate || return 1 # Deactivation should never fail!
+          return "$ret"
+        }
+      fi
+    fi
+
+    # Avoid `error: can't find Rust compiler`.
```

Comment:
What if instead of trying to pinpoint which platforms need cryptography<3.4 we just try to install the latest version and if that fails fallback to cryptography<3.4? Like this:
```
python -m pip install cryptography || python -m pip install 'cryptography<3.4' || ...

python -m pip install -U "${packages[@]}" || ...
```

This is simpler and should work on more platforms.

Reply:
It is indeed simpler, but I deliberately opted for the current approach in order to be _very_ explicit about conditions that require workarounds and as narrow as possible in the application of said workarounds.

This was motivated by the status quo where [generally-applied workarounds](https://github.com/mongodb-labs/drivers-evergreen-tools/blob/4136a431f23461f1abe415e93d85fa55ec4c1be7/.evergreen/csfle/activate_venv.sh#L19) such as pinning cryptography to `~=3.4.8` or using `CRYPTOGRAPHY_DONT_BUILD_RUST=1` continued to demonstrate unexpected failures, and the conditions for said failures appeared to be inconsistent and opaque. It was unclear to me whenever I encountered such a failure whether it was already known, a new problem, or where the blame should be assigned (did _I_ break it, or did the environment change without my knowing?).

My hope was that being explicit in this manner would make it easier to maintain this script moving forward, with simplifications/removals of special-casing being applied in a controlled and targeted manner.

Reply:
I'd prefer the generic one to avoid needing to tweak and maintain these 30 extra lines which may or may not even cover all the hosts. I think a good compromise would be to use the generic approach but add an informative comment that specifically explains why the workaround exists like:
```
# Installing newer versions of cryptography requires rust when a wheel is not available.
# Fallback to an older version that does not require rust if the install fails. This is needed
# for at least the RHEL 6.2, powerpc64le, zSeries, and power8 hosts.
python -m pip install cryptography || python -m pip install 'cryptography<3.4' || ...
```

What do you think?

Reply:
I think that is an acceptable compromise. Would appeciate other reviewers' thoughts on this before committing to the refactor.

Reply:
I have a slight preference for the compromise. That may require less changes to this script as distros undergo changes or more distros are added.

Reply:
Done. Verified by [this patch](https://spruce.mongodb.com/version/6372b59c9ccd4e32ba9533e1).